[1-3단계 - 방탈출 사용자 예약] 밍트(김명지) 미션 제출합니다.

Starlight258 opened the pull request:
제로 안녕하세요 😁  밍트라고 합니다.
[변경 범위](https://github.com/woowacourse/spring-roomescape-member/pull/220/files/527e858b61b44c77c567ef0a313e674bb25af5e2..1eaa69c7e373216a9e0a3705138d47a5c104939b) 입니다.

미션 잘 부탁드립니다! 🔥🔥
감사합니다 🙇‍♀️

<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - [프롤로그 링크](https://prolog.techcourse.co.kr/studylogs/4316) 
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->
미션을 구현하느라 객체지향적으로 개선할 부분을 깊게 생각하지 못했습니다.

## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [ ] 아니오. 전혀 경험이 없습니다.
- [ ] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [ ] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [x] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

## 베이스 코드 선택 체크
- [ ] 이전 미션의 내 코드에서 시작 
- [x] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

기능 구현이 잘 되었는지, 객체지향적으로 개선할 부분이 있다면 피드백 부탁드립니다!

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

solarsalt committed:
docs: add PR template

goohong committed:
feat: setup project

goohong committed:
test: 불필요한 테스트 삭제

goohong committed:
docs: 기능 문서 작성

goohong committed:
feat: 예외 처리 추가

goohong committed:
feat: 커스텀 예외 추가

goohong committed:
feat: 테마 추가

goohong committed:
feat: delete 요청 응답 상태 코드 변경

goohong committed:
feat: 사용자 예약 가능 시간 조회 기능 구현

goohong committed:
feat: 인기 테마 조회 기능 구현

goohong committed:
docs: README 제목 수정

goohong committed:
style: 코드 포맷팅

goohong committed:
fix: 누락한 @PostMapping Annotation 추가

Starlight258 committed:
docs: 상태 코드 수정

asebn1 commented on src/main/java/roomescape/common/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,35 @@
+    @ExceptionHandler(ReservationTimeInUseException.class)
+    public ResponseEntity<String> handleReservationTimeInUseException(ReservationTimeInUseException e) {
+        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(e.getMessage());
+    }
```
아래 두 가지 표기로 사용하고있는데 통일해볼까요?
```java
ResponseEntity.badRequest().body(e.getMessage());
ResponseEntity.status(HttpStatus.BAD_REQUEST).body(e.getMessage());
```

Starlight258 replied:
하나의 표기로 사용하도록 수정했습니다!

[refactor: 동일한 상태 코드 표기로 수정](https://github.com/woowacourse/spring-roomescape-member/pull/220/commits/d1c19cfc683e48ab10b4f9a57e57f2dffceeb634)

asebn1 commented on src/main/java/roomescape/common/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    // TODO : 적합한 상태코드 알아보기
+    @ExceptionHandler(ReservationTimeInUseException.class)
```
TODO 부분을 작성해주셨네요~ 
확인하셨을까요?

Starlight258 replied:
TODO 모두 해결하여 반영하였습니다!

[docs: todo: not found body 포함 여부 주석 삭제](https://github.com/woowacourse/spring-roomescape-member/pull/220/commits/d0dfd74e33af124afc61a19e6d152cecb641b576)

[fix: 리소스 생성시 location 헤더에 적절한 uri 반환하도록 수정](https://github.com/woowacourse/spring-roomescape-member/pull/220/commits/10c1f4d7bd5817413b979540808af5492f26940b)

[feat: main,test db resource 분리](https://github.com/woowacourse/spring-roomescape-member/pull/220/commits/f5f6d78583c6d3bd56dc03a9bb877c6474822774)

## Created URI
`ReservationTimeController`
```java
    @PostMapping
    public ResponseEntity<ReservationTimeResponse> createReservationTime(
            @RequestBody ReservationTimeCreateRequest request
    ) {
        ReservationTimeResponse dto = reservationTimeService.create(request);
        return ResponseEntity.created(makeReservationTimeDto(dto)).body(dto);
    }

    private URI makeReservationTimeDto(final ReservationTimeResponse dto) {
        return ServletUriComponentsBuilder
                .fromCurrentRequest()
                .path("/{id}")
                .buildAndExpand(dto.id())
                .toUri();
    }
```
예약 시간 생성시 Created(201) 응답과 함께 생성된 리소스의 위치인 uri를 포함하였습니다.
그런데 현재는 예약 전체 조회 api(`/reservations`)만 존재하고 단건 예약 조회 api는 존재하지 않는데, 위와 같이 존재하지 않은 uri를 명시해도 되는지 궁금합니다.
따로 예약 단건 조회 api를 생성하는 것이 좋을까요? 🤔

asebn1 replied:
단건 조회가 필요한 스펙이 아니라면, 201 Created만 반환하고 Location은 생략하는 쪽이 더 자연스러운 설계일 것 같아요.
반대로 해당 URI를 클라이언트가 추후 사용할 수 있도록 할 계획이라면, 단건 조회 API를 함께 제공해주는 것도 좋은 방향이라고 생각합니다.

Starlight258 replied:
존재하지 않은 URI라면 명시하지 않는 것이 좋을 것 같습니다!
[feat: 존재하지 않은 URI 응답에 포함되지 않도록 제거](https://github.com/Starlight258/spring-roomescape-member/commit/565fb9d6207c4b606c9a445aa20aa4dfe49ca1ca)

asebn1 commented on src/main/java/roomescape/reservation/controller/ReservationController.java:
```diff
@@ -27,18 +28,20 @@ public ResponseEntity<List<ReservationResponse>> getReservations() {
     ) {
         reservationService.delete(id);
-        return ResponseEntity.ok().build();
+        return ResponseEntity.noContent().build();
```
👍👍

asebn1 commented on src/main/java/roomescape/reservation/controller/ReservationController.java:
```diff
@@ -27,18 +28,20 @@ public ResponseEntity<List<ReservationResponse>> getReservations() {
             @RequestBody ReservationCreateRequest request
     ) {
-        return ResponseEntity.ok(reservationService.create(request));
+        URI location = URI.create("http://localhost:8080");
```
`http://localhost:8080`가 아니라 생성된 ID나 의미있는 값을 넣어볼까요?

Starlight258 replied:
네! 위에서 언급해주신대로 Uri 수정하였습니다!

asebn1 commented on src/main/java/roomescape/reservation/controller/ReservationController.java:
```diff
@@ -27,18 +28,20 @@ public ResponseEntity<List<ReservationResponse>> getReservations() {
             @RequestBody ReservationCreateRequest request
     ) {
-        return ResponseEntity.ok(reservationService.create(request));
+        URI location = URI.create("http://localhost:8080");
```
`http://localhost:8080`가 아닌 생성된 ID나 의미있는 값을 넣어볼 수 있을까요?

Starlight258 replied:
네! 위에서 언급해주신대로 Uri 수정하였습니다!

asebn1 commented on src/main/java/roomescape/reservation/domain/Reservation.java:
```diff
@@ -35,6 +48,14 @@ private void validateDateTime(LocalDate date, LocalTime time) {
+        return id.getValue();
+    }
+
+    public void setId(Long value) {
```
불필요한 setter 같아요

Starlight258 replied:
사용하지 않는 setter는 제거하겠습니다!

[feat: 필요없는 setter 제거](https://github.com/woowacourse/spring-roomescape-member/pull/220/commits/19491f66afcf1deb0b7223ac96e535a4cdaa9497)

asebn1 commented on src/main/java/roomescape/reservation/dto/response/ReservationResponse.java:
```diff
@@ -4,16 +4,19 @@
+        ThemeResponse theme
 ) {
-    public static ReservationResponse from(Reservation reservation, ReservationTime reservationTime) {
+    public static ReservationResponse from(Reservation reservation, ReservationTime reservationTime, Theme theme) {
```
매개변수가 여러 개라면 from이 아닌 of를 사용합니다.

- [정적 팩터리 메서드 명명 방식](https://velog.io/@lychee/%EC%9D%B4%ED%8E%99%ED%8B%B0%EB%B8%8C-%EC%9E%90%EB%B0%94-%EC%95%84%EC%9D%B4%ED%85%9C-1.-%EC%83%9D%EC%84%B1%EC%9E%90-%EB%8C%80%EC%8B%A0-%EC%A0%95%EC%A0%81-%ED%8C%A9%ED%84%B0%EB%A6%AC-%EB%A9%94%EC%84%9C%EB%93%9C%EB%A5%BC-%EA%B3%A0%EB%A0%A4%ED%95%98%EB%9D%BC) 참고

Starlight258 replied:
수정했습니다! 페어의 코드에서도 리팩토링할 부분이 있는지 확인해보겠습니다. 감사합니다!

[feat: 정적 팩토리 메서드에서 파라미터가 여러개일 때 from->of로 수정](https://github.com/woowacourse/spring-roomescape-member/pull/220/commits/bfd3a8400e4264803e8f55115cd25e449ed9c128)

asebn1 commented on src/main/java/roomescape/reservation/dto/request/ReservationCreateRequest.java:
```diff
@@ -1,10 +1,18 @@
+        long themeId
 ) {
+    public ReservationCreateRequest {
+        if (name == null) {
```
이렇게 하면 `" "` 공백은 통과할 것 같아요
`name.isBlank()`를 추가하면 어떨까요?

Starlight258 replied:
문자열 검증에서는 null 뿐 아니라 비어있는지도 함께 확인하는 것이 좋을 것 같습니다! 감사합니다.

[feat: 비어있는지를 확인하는 문자열 검증 추가](https://github.com/woowacourse/spring-roomescape-member/pull/220/commits/712111a47e1133b808dc6c3abf99c80d981edf9a)

asebn1 commented on src/main/java/roomescape/reservation/repository/JDBCReservationRepository.java:
```diff
@@ -55,14 +68,85 @@ public Reservation put(final Reservation reservation) {
+                }, date, themeId);
+    }
+
 }
```
```suggestion
    }
}
```
개행 컨밴션도 지켜보아요 :)

Starlight258 replied:
수정하였습니다! 
[style: 포맷팅](https://github.com/woowacourse/spring-roomescape-member/pull/220/commits/cd7727e9db211756cd5906b622c27dcddd826cdc)

포맷팅을 하면서 궁금한 점이 있어 질문드립니다! 제로는 새롭게 클래스를 만들고 나서 한줄을 띄우시나요?
```java
@Repository
public class JDBCReservationRepository implements ReservationRepository {
    // 여기 부분
    private final JdbcTemplate jdbcTemplate;

    public JDBCReservationRepository(final JdbcTemplate jdbcTemplate) {
        this.jdbcTemplate = jdbcTemplate;
    }


```
전 항상 한 줄을 띄우는데, 페어는 붙여 쓰더라구요. 개발자마다 다른 것 같은데, 제로는 어떤 편이신지 궁금해요!

asebn1 replied:
저도 한 줄 띄우는 편입니다.
정답은 없다고 생각해요 ㅎㅎ 팀 컨밴션에 맞추는게 좋다고 생각합니다.

asebn1 commented on src/main/java/roomescape/reservation/repository/JDBCReservationRepository.java:
```diff
@@ -55,14 +68,85 @@ public Reservation put(final Reservation reservation) {
+                        + "on r.time_id = t.id "
+                        + "inner join theme as th "
+                        + "on r.theme_id = th.id "
+                        + "where (date, theme_id) = (?, ?)",
```
`where (date, theme_id) = (?, ?)` 같이도 쓸 수 있었군요 👀
저도 하나 알아갑니다 ㅎㅎ

asebn1 commented on src/main/java/roomescape/reservation/repository/JDBCReservationRepository.java:
```diff
@@ -55,14 +68,85 @@ public Reservation put(final Reservation reservation) {
+        ));
+    }
+
+    @Override
+    public boolean existsByDateAndTimeId(final LocalDate date, final long timeId) {
+        return Boolean.TRUE.equals(jdbcTemplate.queryForObject(
+                "SELECT EXISTS (SELECT 1 FROM reservation WHERE (date, time_id) = (?,?))",
+                Boolean.class,
+                date,
+                timeId
+        ));
+    }
+
+    @Override
+    public List<Reservation> findByDateAndThemeId(final LocalDate date, final Long themeId) {
```
- `Long themeId`
- `long timeId`

일관되게 맞춰줍시다!

Starlight258 replied:
수정했습니다!

[feat: id에 대한 타입을 모두 Long으로 통일](https://github.com/woowacourse/spring-roomescape-member/pull/220/commits/19f9145a4124346261d5735546600b2c5b37a20e)

asebn1 commented on src/main/java/roomescape/reservationtime/controller/ReservationTimeController.java:
```diff
@@ -24,19 +28,27 @@ public ReservationTimeController(final ReservationTimeService reservationTimeSer
 
     @DeleteMapping("/{id}")
-    public ResponseEntity<Void> deleteReservations(
+    public ResponseEntity<Void> deleteReservationTimes(
```
여기도 no Content를 적용해봅시다~

Starlight258 replied:
반영했습니다! 감사합니다.

asebn1 commented on src/main/java/roomescape/reservationtime/dto/response/AvailableReservationTimeResponse.java:
```diff
@@ -0,0 +1,18 @@
+        LocalTime startAt,
+        boolean alreadyBooked
+) {
+    public static AvailableReservationTimeResponse from(ReservationTime reservationTime, boolean alreadyBooked) {
```
```suggestion
    public static AvailableReservationTimeResponse of(ReservationTime reservationTime, boolean alreadyBooked) {
```
여기도 맞출 수 있겠네요!

Starlight258 replied:
위에서 언급하신대로 dto의 생성자에서 파라미터가 여러개인 생성자에 대해 모두 수정하였습니다 :)

asebn1 commented on src/main/java/roomescape/reservationtime/repository/JDBCReservationTimeRepository.java:
```diff
@@ -4,15 +4,13 @@
 import roomescape.reservationtime.entity.ReservationTimeEntity;
 
 @Repository
-@Primary
```
- `@Primary`란 무엇일까요?
- 이전에 추가했던 이유가 있을까요?

Starlight258 replied:
@Primary는 같은 유형의 빈이 여러개 있을 때, 해당 빈을 사용하도록 우선순위를 높이는 어노테이션입니다.
기존에는 `ReservationTimeRepository`의 구현체가 `Memory`와 `Jdbc` 둘 다 있어 `Jdbc` 구현체만 사용하기 위해 붙여주었습니다!

asebn1 commented on src/main/java/roomescape/reservationtime/controller/ReservationTimeController.java:
```diff
@@ -24,19 +28,27 @@ public ReservationTimeController(final ReservationTimeService reservationTimeSer
     @DeleteMapping("/{id}")
-    public ResponseEntity<Void> deleteReservations(
+    public ResponseEntity<Void> deleteReservationTimes(
             @PathVariable("id") long id
```
`Long id`도 있고 `long id` 있네요

Starlight258 replied:
넵 위에서 언급하신것처럼, id는 `Long`을 사용하도록 일괄 수정하였습니다 :) 감사합니다!

asebn1 commented on src/main/java/roomescape/reservationtime/service/ReservationTimeService.java:
```diff
@@ -36,7 +52,32 @@ public ReservationTimeResponse create(final ReservationTimeCreateRequest request
+        List<Reservation> reservations = reservationRepository.findByDateAndThemeId(date, themeId);
+        Map<ReservationTime, Boolean> availabilities = new HashMap<>();
+
+        for (ReservationTime reservationTime : reservationTimes) {
+            Long timeId = reservationTime.getId();
+            availabilities.put(reservationTime, false);
+
+            for (Reservation reservation : reservations) {
+                if (Objects.equals(timeId, reservation.getTime().getId())) {
+                    availabilities.put(reservationTime, true);
+                }
+            }
+        }
```
이중 포문을 사용하고 있네요 
개선할 수 있을까요?

Starlight258 replied:
우선 이중 `for` 문을 `stream`으로 변경하였습니다.
[refactor: 이중 for문 stream으로 변경](https://github.com/woowacourse/spring-roomescape-member/pull/220/commits/f962cf2317c701668af0c441577ddce1966747fb)

### 쿼리 vs 서비스 메서드
궁금한 점이 있습니다. 현재는 날짜와 테마에 해당하는 예약을 조회하고(`findByDateAndThemeId()`), 모든 예약 시간들을 조회해서(`getAll()`) 예약이 존재하는 예약 시간과 예약이 존재하지 않은 예약 시간을 서비스에서 for문으로 돌면서 처리하고 있습니다.

```java
 public List<AvailableReservationTimeResponse> getAvailableReservationTimes(final LocalDate date,
                                                                               final Long themeId) {
        List<ReservationTime> reservationTimes = reservationTimeRepository.getAll();
        List<Reservation> reservations = reservationRepository.findByDateAndThemeId(date, themeId);

        Map<ReservationTime, Boolean> availabilities = reservationTimes.stream()
                .map(reservationTime -> mapToDto(reservations, reservationTime))
                .collect(Collectors.toMap(Entry::getKey, Entry::getValue));

        return availabilities.entrySet().stream()
                .map(entry -> {
                    ReservationTime reservationTime = entry.getKey();
                    return new AvailableReservationTimeResponse(reservationTime.getId(), reservationTime.getStartAt(),
                            entry.getValue());
                })
                .toList();
    }
```
서비스에서 처리하지 않고 db에서 쿼리로 한번에 가져오는 방법이 있을 것 같은데, 그렇게 되면 쿼리가 너무 복잡해질 것 같습니다. 하지만, db 관점에서 좀 더 최적화해서 조회하여 가져와 성능상 좋을 것 같습니다 (추측입니다.)

제로는 **서비스에서 가독성 있게 for문을 돌면서 결과값을 생성**하는 것이 좋다고 생각하시나요, **db 쿼리로 한번에 가져오고 서비스에서는 단순 조회한 데이터를 dto로 생성하는 역할**만 수행하는 것이 좋다고 생각하시는지 궁금합니다!

---
추가적으로 이번 미션을 하면서 쿼리 작성에 어려움을 많이 느꼈는데요 🥹, **쿼리 작성 실력이나 db 다루는 법**을 어떻게 늘릴 수 있는지 제로의 팁이 있으면 알려주시면 큰 도움이 될 것 같습니다!

asebn1 replied:
> 제로는 서비스에서 가독성 있게 for문을 돌면서 결과값을 생성하는 것이 좋다고 생각하시나요, db 쿼리로 한번에 가져오고 서비스에서는 단순 조회한 데이터를 dto로 생성하는 역할만 수행하는 것이 좋다고 생각하시는지 궁금합니다!

- 성능 부담이 도메인 복잡도보다 크다면, 쿼리로 최대한 해결을 시도해볼 것 같고
- 반대로 도메인 복잡도가 더 크다면, 서비스 단에서 처리하는 편이 더 유지보수에 유리하다고 생각합니다.
요구사항에 맞추어 유연하게 작성하는 것이 좋다고 생각합니다 :)

> 추가적으로 이번 미션을 하면서 쿼리 작성에 어려움을 많이 느꼈는데요 🥹, 쿼리 작성 실력이나 db 다루는 법을 어떻게 늘릴 수 있는지 제로의 팁이 있으면 알려주시면 큰 도움이 될 것 같습니다!

저도 우테코에서 계속 고민했던 부분인데요!
저는 Real MySQL 8.0을 가볍게 1회독하면서 관련 개념을 먼저 익혔고,
이후 레벨3 팀 프로젝트에서 DB 관련 구조나 인덱스 설계 등을 제가 먼저 제안하고 맡아보는 방식으로 연습해 보았습니다.
프로젝트 안에서 "어떤 점을 적용할 수 있을까"를 고민한 것이 가장 도움이 된 것 같아요.
팀 내에서 DB 구조나 성능 개선과 관련된 역할을 맡아보시는 걸 추천드립니다 ㅎㅎ

asebn1 commented on src/main/java/roomescape/theme/domain/Theme.java:
```diff
@@ -0,0 +1,45 @@
+        return id.getValue();
+    }
+
+    public void setId(final Long value) {
```
제거해봅시다

Starlight258 replied:
위에서 언급하신대로 사용하지 않은 setter는 모두 제거하였습니다 :) 감사합니다.

asebn1 commented on src/main/java/roomescape/theme/domain/Theme.java:
```diff
@@ -0,0 +1,45 @@
+    private final String description;
+    private final String thumbnail;
+
+    public Theme(final Id id, final String name, final String description, final String thumbnail) {
```
Theme에 대한 필드 유효성 검증은 없을까요?

Starlight258 replied:
`Theme`에 대한 `name`, `description`, `thumbnail` 검증을 추가하였습니다 :) 

[feat: Theme 검증 추가](https://github.com/woowacourse/spring-roomescape-member/pull/220/commits/ab768bf7b3adbb23a120000c0bc849150935ccb5)

asebn1 commented on src/main/java/roomescape/theme/repository/JDBCThemeRepository.java:
```diff
@@ -0,0 +1,95 @@
+        return jdbcTemplate.update("DELETE FROM theme WHERE id = ?", id) != 0;
+    }
+
+    @Override
+    public Optional<Theme> findById(final long id) {
+        ThemeEntity themeEntity = jdbcTemplate.queryForObject(
+                "SELECT id, name, description, thumbnail FROM theme WHERE id = ?",
+                (resultSet, rowNum) -> new ThemeEntity(
+                        resultSet.getLong("id"),
+                        resultSet.getString("name"),
+                        resultSet.getString("description"),
+                        resultSet.getString("thumbnail")
+                ), id
+        );
+        return Optional.ofNullable(themeEntity)
+                .map(ThemeEntity::toTheme);
+    }
```
위와같이 `queryForObject` 부분도 챙겨주세요!

Starlight258 replied:
[fix: queryForObject에서 빈 결과 조회시 예외 발생하지 않도록 수정](https://github.com/Starlight258/spring-roomescape-member/commit/17f58761cb1abbfd079a08b4d72ab001e71c92bc)

asebn1 reviewed (changes requested):
밍트 안녕하세요~
이번 미션 리뷰어 제로입니다
몇 가지 코멘트 남겼습니다
확인해주세요 :)

Starlight258 committed:
refactor: 동일한 상태 코드 표기로 수정

Starlight258 committed:
feat: 적절한 상태 코드로 수정

Starlight258 committed:
docs: todo: not found body 포함 여부 주석 삭제

- not found에서 body를 포함하여 오류 메세지를 제공하는 것은 선택이다. 따라서 계속 오류 메세지를 제공하도록 한다.

Starlight258 committed:
fix: 리소스 생성시 location 헤더에 적절한 uri 반환하도록 수정

Starlight258 committed:
feat: main,test db resource 분리

- 테스트용 설정 파일(application.yml)을 생성한다.
- 프로덕션과 test 모두 schema.sql을 공통으로 사용하므로 test용 schema.sql은 따로 작성하지 않는다.

Starlight258 committed:
feat: 리소스 삭제시 noContent 상태코드 반환하도록 수정

Starlight258 committed:
feat: 필요없는 setter 제거

Starlight258 committed:
feat: 정적 팩토리 메서드에서 파라미터가 여러개일 때 from->of로 수정

Starlight258 committed:
feat: 비어있는지를 확인하는 문자열 검증 추가

Starlight258 committed:
style: 포맷팅

Starlight258 committed:
test: test data repository가 아닌 service에서 설정하도록 수정

- repository를 사용할 경우 적절한 비즈니스 로직이 수행되지 않고 바로 db에 저장되므로 위험하다.

Starlight258 commented on src/test/java/roomescape/reservation/service/ReservationServiceTest.java:
```diff
@@ -0,0 +1,95 @@
+class ReservationServiceTest {
+    private final LocalDate futureDate = LocalDate.now().plusDays(1);
+
+    private ReservationService reservationService;
+    private ReservationTimeRepository reservationTimeRepository;
+
```
## Service test data : `Service` vs `Repository`
### `Service`에서 `test data` 설정
`service`를 이용해서 테스트 데이터를 설정하면 `dto`를 생성하여 요청을 보내고, `repository`에 대한 의존성이 사라집니다.
하지만 만약 자기 자신의 메서드를 사용한다면 해당 메서드가 유효하지 않은 경우 연속으로 테스트가 깨집니다.

### `repository`
 `repository`에서 테스트 데이터를 설정하면 테스트 대상인 `service`의 메서드를 사용하지 않으므로 service의 메서드가 잘못되어도 테스트가 연속해서 깨지지 않습니다. 

하지만 의존성 주입 부분이 늘어나고, **service에서 수행한 비즈니스 로직을 수행하지 않고 바로 repository에 저장하므로 유효하지 않은 데이터가 저장되는 문제가 있습니다.**

현재는 `service`와 `repository`를 함께 사용하고 있는데, 하나로 정해서 테스트 데이터를 설정하는 것이 좋을 것 같습니다.

저는 `service`를 사용하는 것이 기능에 대한 유효한 테스트 데이터를 생성하므로 좋다고 생각하는데, 제로의 생각이 궁금해요!

asebn1 replied:
저도 가능한 경우엔 Service를 통해 테스트 데이터를 생성하는 쪽을 선호합니다.
비즈니스 로직을 거친 유효한 데이터가 보장되기 때문이에요.
밍트가 이야기 주신것처럼 Repository로 직접 주입하면 테스트의 독립성은 확보되지만 실제 상황과는 다른 비정상적인 데이터가 들어갈 수 있어 주의가 필요할 수 있겠네요 ㅎㅎ

Starlight258 committed:
test: 참조하지 않은 필드 로컬 변수로 수정

Starlight258 committed:
feat: id에 대한 타입을 모두 Long으로 통일

Starlight258 committed:
refactor: 이중 for문 stream으로 변경

Starlight258 committed:
feat: Theme 검증 추가

Starlight258 commented:
![image](https://github.com/user-attachments/assets/5c2ddc33-6d1b-4688-9076-794c37c0514b)

이 리뷰에 대해 좀 더 구체적으로 말씀해주실 수 있나요? 어떤 부분을 수정하는 것이 좋을까요?
해당 글에 댓글을 작성할 수 없어 부득이하게 캡쳐하였습니다!

Starlight258 commented:
@asebn1 리뷰 감사합니다 제로 :)
많은 부분에서 빼먹은 부분이 많은데, 잘 챙겨주셔서 감사합니다 😁
개선할 부분 있으면 피드백 부탁드립니다. 🔥🔥
감사합니다 🙇‍♀️

Starlight258 committed:
feat: @RestController, @RestControllerAdvice 사용

- rest api를 제공하는 application이므로 의미적으로 명확하게 표현하기 위해 @RestController, @RestControllerAdvice를 사용한다.

Starlight258 committed:
fix: view를 반환하는 controller는 @Controller 사용하도록 변경

asebn1 commented:
queryForObject는 [coderabbitai 코멘트](https://github.com/woowacourse/spring-roomescape-member/pull/220#discussion_r2071122896) 아래에 단 내용이었어요!

![image](https://github.com/user-attachments/assets/79976d97-9f5d-4407-92f8-97dfd51fd110)

asebn1 reviewed (approved):
밍트 안녕하세요~
이번 단계 잘 마무리해주셔서 
이만 머지하겠습니다
다음 단계도 화이팅! 🔥

Starlight258 commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,49 @@
+- [ ] 이전 미션의 내 코드에서 시작 
+- [ ] 이전 미션의 페어의 코드에서 시작
+
+## 어떤 부분에 집중하여 리뷰해야 할까요?
+<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
+리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
+반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->
+
```
-- front code --

Starlight258 commented on src/main/resources/templates/signup.html:
```diff
@@ -1,65 +1,65 @@
+            <input class="form-control" id="password" placeholder="Enter password" type="password">
+        </div>
+        <div class="form-group">
+            <label for="name">Name</label>
+            <input class="form-control" id="name" placeholder="Enter name" type="name">
```
-- front code --

Starlight258 commented on src/main/resources/templates/reservation.html:
```diff
@@ -1,95 +1,95 @@
+    <div class="button-group float-right">
+        <button id="reserve-button" class="btn btn-primary mt-3 disabled">예약하기</button>
+    </div>
 </div>
 </div>
```
-- front code --

Starlight258 commented on src/main/resources/static/js/reservation-with-member.js:
```diff
@@ -8,239 +8,239 @@ const themesOptions = [];
-  button.textContent = label;
-  button.classList.add('btn', className, 'mr-2');
-  button.addEventListener('click', eventListener);
-  return button;
+    const button = document.createElement('button');
+    button.textContent = label;
+    button.classList.add('btn', className, 'mr-2');
+    button.addEventListener('click', eventListener);
+    return button;
 }
```
-- front code --

Starlight258 commented on src/main/resources/static/js/reservation-with-member.js:
```diff
@@ -8,239 +8,239 @@ const themesOptions = [];
-        if (response.status === 201) return response.json();
-        throw new Error('Create failed');
-      });
+    return fetch('/admin/reservations', requestOptions)
```
[feat: api endpoint 수정](https://github.com/Starlight258/spring-roomescape-member/commit/8dfe6721f19eeb083f248d08e7ec49e6e27a15ae)

Starlight258 commented on src/main/resources/static/js/reservation-with-member.js:
```diff
@@ -8,239 +8,239 @@ const themesOptions = [];
+    TODO: [6단계] 예약 검색 - 조건에 따른 예약 조회 API 호출
+          요청 포맷에 맞게 설정
+    */
+    fetch('/', { // 예약 검색 API 호출
+        method: 'GET',
+        headers: {
+            'Content-Type': 'application/json'
+        },
+    }).then(response => {
+        if (response.status === 200) return response.json();
+        throw new Error('Read failed');
+    }).then(render)
+        .catch(error => console.error("Error fetching available times:", error));
```
-- front code --

Starlight258 commented on src/main/resources/static/js/user-scripts.js:
```diff
@@ -1,152 +1,152 @@
-        console.error('Error during login:', error);
-      });
+        .then(response => {
+            if (200 === !response.status) {
```
-- front code --

Starlight258 commented on src/main/resources/static/js/reservation-new.js:
```diff
@@ -6,193 +6,193 @@ const timesOptions = [];
-  button.textContent = label;
-  button.classList.add('btn', className, 'mr-2');
-  button.addEventListener('click', eventListener);
-  return button;
+    const button = document.createElement('button');
+    button.textContent = label;
+    button.classList.add('btn', className, 'mr-2');
+    button.addEventListener('click', eventListener);
+    return button;
 }
```
-- front code --

Starlight258 commented on src/main/resources/static/js/time.js:
```diff
@@ -2,133 +2,134 @@ let isEditing = false;
-  const input = document.createElement('input');
-  input.type = 'time'
-  input.className = 'form-control';
-  return input;
+    const input = document.createElement('input');
+    input.type = 'time'
+    input.className = 'form-control';
+    return input;
 }
```
-- front code --

Starlight258 commented on src/main/resources/static/js/time.js:
```diff
@@ -2,133 +2,134 @@ let isEditing = false;
+    const body = createBody(inputs);
 
-  requestCreate(body)
-      .then(() => {
-        location.reload();
-      })
-      .catch(error => console.error('Error:', error));
+    requestCreate(body)
+        .then(() => {
+            location.reload();
+        })
+        .catch(error => console.error('Error:', error));
 
-  isEditing = false;  // isEditing 값을 false로 설정
+    isEditing = false;  // isEditing 값을 false로 설정
 }
```
-- front code --

Starlight258 commented on src/main/resources/templates/admin/reservation.html:
```diff
@@ -1,58 +1,58 @@
+    <div class="table-header">
+        <button class="btn btn-custom mb-2 float-right" id="add-button">예약 추가</button>
+    </div>
+    <div class="table-container"/>
```
-- front code --

Starlight258 commented on src/main/resources/static/js/time.js:
```diff
@@ -2,133 +2,134 @@ let isEditing = false;
-      .then(response => {
-        if (response.status === 201) return response.json();
-        throw new Error('Create failed');
-      });
+    const requestOptions = {
+        method: 'POST',
+        headers: {'Content-Type': 'application/json'},
+        body: JSON.stringify(data)
+    };
+
+    return fetch(API_ENDPOINT, requestOptions)
+        .then(response => {
+            if (response.status === 201) return response.json();
+            throw new Error('Create failed');
+        });
 }
```
-- front code --

Starlight258 commented on src/main/resources/static/js/reservation-new.js:
```diff
@@ -6,193 +6,193 @@ const timesOptions = [];
+    const timeSelect = row.querySelector('#time-select');
+
+    const reservation = {
+        name: nameInput.value,
+        themeId: themeSelect.value,
+        date: dateInput.value,
+        timeId: timeSelect.value
+    };
```
-- front code --

Starlight258 commented on src/main/java/roomescape/theme/dto/request/ThemeCreateRequest.java:
```diff
@@ -0,0 +1,13 @@
+
+import roomescape.theme.domain.Theme;
+
+public record ThemeCreateRequest(
+        String name,
+        String description,
+        String thumbnail
+) {
+    public Theme toTheme() {
+        return Theme.withUnassignedId(name, description, thumbnail);
+    }
+}
```
[feat: dto 검증 코드 추가](https://github.com/Starlight258/spring-roomescape-member/commit/09c7e03c4d8b5b0b6766bb24c2d517cb1adefb3d)

Starlight258 commented on src/test/java/roomescape/theme/repository/JDBCThemeRepositoryTest.java:
```diff
@@ -0,0 +1,33 @@
+        themeRepository = new JDBCThemeRepository(jdbcTemplate);
+    }
+
+    @Test
+    void findTop10PopularThemesWithinLastWeek_shouldReturnCorrectly() {
+        List<Theme> top10PopularThemesWithinLastWeek = themeRepository.findTop10PopularThemesWithinLastWeek(futureDate);
+        for (Theme theme : top10PopularThemesWithinLastWeek) {
+            System.out.println(theme.getId());
+        }
+    }
```
[test: 인기 테마 조회 테스트 추가](https://github.com/Starlight258/spring-roomescape-member/commit/b57b56ca14dc62fee3ad0a2287d895d0b248c1aa)

Starlight258 commented on src/main/java/roomescape/reservationtime/service/ReservationTimeService.java:
```diff
@@ -0,0 +1,83 @@
+                            entry.getValue());
+                })
+                .toList();
+    }
```
[refactor: 인기 테마 조회 성능 최적화](https://github.com/Starlight258/spring-roomescape-member/commit/f83395979cfbd83ebc606ec4a2f38d4fa1a3aff1)

Starlight258 commented on src/main/java/roomescape/reservationtime/domain/ReservationTime.java:
```diff
@@ -0,0 +1,58 @@
+        this.startAt = startAt;
+    }
+
+    private ReservationTime(final Id id, final String startAt) {
+        this.id = id;
+        this.startAt = LocalTime.parse(startAt);
+    }
```
globalExceptionHandler로 이미 처리

Starlight258 commented on src/main/resources/static/js/user-reservation.js:
```diff
@@ -1,202 +1,202 @@
+
+    if (selectedDate && selectedThemeId && selectedTimeId) {
+
+        /*
+        TODO: [3단계] 사용자 예약 - 예약 요청 API 호출
+              [5단계] 예약 생성 기능 변경 - 사용자
+              request 명세에 맞게 설정
+        */
+        const reservationData = {
+            date: selectedDate,
+            themeId: selectedThemeId,
+            timeId: selectedTimeId,
+            name: name
+        };
```
[docs: 수행한 todo 제거](https://github.com/Starlight258/spring-roomescape-member/commit/08edc4cb675da926eece8d64ffcd9e4be5eb6e56)

Starlight258 commented on src/main/java/roomescape/theme/controller/ThemeController.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    @PostMapping
+    public ResponseEntity<ThemeResponse> createTheme(
+            @RequestBody ThemeCreateRequest request
+    ) {
```
validation 라이브러리 사용하지 않고 생성자 & 도메인에서 처리

Starlight258 commented on src/main/java/roomescape/theme/controller/ThemeController.java:
```diff
@@ -0,0 +1,51 @@
+        return ResponseEntity.ok(themeService.getPopularThemes());
+    }
+
+    @PostMapping
+    public ResponseEntity<ThemeResponse> createTheme(
+            @RequestBody ThemeCreateRequest request
+    ) {
+        return ResponseEntity.created(URI.create("")).body(themeService.create(request));
+    }
```
해결

Starlight258 commented on src/main/java/roomescape/reservationtime/repository/JDBCReservationTimeRepository.java:
```diff
@@ -0,0 +1,75 @@
+        );
+        return Optional.ofNullable(reservationTimeEntity)
+                .map(ReservationTimeEntity::toReservationTime);
+    }
```
[fix: queryForObject에서 빈 결과 조회시 예외 발생하지 않도록 수정](https://github.com/Starlight258/spring-roomescape-member/commit/17f58761cb1abbfd079a08b4d72ab001e71c92bc)

Starlight258 commented on src/main/java/roomescape/reservation/repository/JDBCReservationRepository.java:
```diff
@@ -0,0 +1,152 @@
+    @Override
+    public List<Reservation> getAll() {
+        return jdbcTemplate.query(
+                "SELECT "
+                        + "r.id as reservation_id, "
+                        + "r.name, "
+                        + "r.date, "
+                        + "t.id as time_id, "
+                        + "t.start_at as time_value, "
+                        + "th.id as theme_id, "
+                        + "th.name as theme_name, "
+                        + "th.description,"
+                        + "th.thumbnail "
+                        + "FROM reservation as r "
+                        + "inner join reservation_time as t "
+                        + "on r.time_id = t.id "
+                        + "inner join theme as th "
+                        + "on r.theme_id = th.id",
+                (resultSet, rowNum) -> {
+                    ReservationTimeEntity timeEntity = new ReservationTimeEntity(
+                            resultSet.getLong("time_id"),
+                            resultSet.getString("time_value"));
+
+                    ThemeEntity themeEntity = new ThemeEntity(
+                            resultSet.getLong("theme_id"),
+                            resultSet.getString("theme_name"),
+                            resultSet.getString("description"),
+                            resultSet.getString("thumbnail"));
+
+                    ReservationEntity entity = new ReservationEntity(
+                            resultSet.getLong("id"),
+                            resultSet.getString("name"),
+                            resultSet.getString("date"),
+                            timeEntity,
+                            themeEntity
+                    );
+                    return entity.toReservation();
+                }
+        );
+    }
```
[refactor: 중복 코드 제거](https://github.com/Starlight258/spring-roomescape-member/commit/456da0671948d838171bee6d1ec5a72c61710e85)

Starlight258 commented on src/main/java/roomescape/reservation/repository/JDBCReservationRepository.java:
```diff
@@ -0,0 +1,152 @@
+                            resultSet.getString("description"),
+                            resultSet.getString("thumbnail"));
+
+                    ReservationEntity entity = new ReservationEntity(
+                            resultSet.getLong("id"),
+                            resultSet.getString("name"),
```
[fix: 별칭 사용하도록 수정](https://github.com/Starlight258/spring-roomescape-member/commit/75d42560cc12b73e0c0cb7db2e7d884fa8c611e8)

Starlight258 commented on src/test/java/roomescape/reservationtime/service/ReservationTimeServiceTest.java:
```diff
@@ -0,0 +1,122 @@
+
+        assertThatThrownBy(() -> reservationTimeService.delete(reservationTime.getId()))
+                .hasMessage("해당 시간에 대한 예약이 존재하여 삭제할 수 없습니다.");
+    }
```
[test: 테스트 코드 예외 클래스까지 검증하도록 수정](https://github.com/Starlight258/spring-roomescape-member/commit/3f06980d24eee63ec0b4388d74c43ca4840f1b9e)
